File: summer2019/week5/leaf-similar-trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> sequence1 = new ArrayList<>();
        leafValueSequence(root1, sequence1);
        List<Integer> sequence2 = new ArrayList<>();
        leafValueSequence(root2, sequence2);
        return sequence1.equals(sequence2);
    }
    
    public void leafValueSequence(TreeNode root, List<Integer> sequence) {
        if (root.left == null && root.right == null) {
            sequence.add(root.val);
        }
        if (root.left != null) {
            leafValueSequence(root.left, sequence);
        }
        if (root.right != null) {
            leafValueSequence(root.right, sequence);
        }
    }
}

File: summer2021/unit5/leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        return self.getLeaves(root1) == self.getLeaves(root2)
    
    def getLeaves(self, root) -> List[int]:
        if not root:
            return []
        leftSide = self.getLeaves(root.left)
        currentVal = []
        if not root.left and not root.right:
            currentVal.append(root.val)
        rightSide = self.getLeaves(root.right)
        return leftSide + currentVal + rightSide